pkg/agent/openflow/feature.go

```diff
@@ -0,0 +1,280 @@
+	pipelineRoot binding.PipelineID = iota
+	// pipelineARP is used to process ARP packets.
+	pipelineARP
+	// pipelineIP is used to process IPv4/IPv6 packets
+	pipelineIP
+	// pipelineMulticast is used to process multicast packets.
+	pipelineMulticast
+)
+
+const (
+	// Stage is used to group tables which implement similar functions in a pipeline. At this moment, we have the following
+	// stages:
+
+	// stageStart is only used to initialize PipelineClassifierTable.
+	stageStart binding.StageID = iota
+	// Classify packets "category" (tunnel, local gateway or local Pod, etc).
+	stageClassifier
+	// Validate packets.
+	stageValidation
+	// Transform committed packets in CT zones.
+	stageConntrackState
+	// Similar to PREROUTING chain of nat table in iptables. DNAT for Service connections is performed in this stage.
+	stagePreRouting
+	// Install egress rules for K8s NetworkPolicy and Antrea NetworkPolicy.
+	stageEgressSecurity
+	// L3 Forwarding of packets.
+	stageRouting
+	// Similar to POSTROUTING chain of nat table in iptables. SNAT for Service connections is performed in this stage.
+	stagePostRouting
+	// L2 Forwarding of packets.
+	stageSwitching
+	// Install ingress rules for K8s NetworkPolicy and Antrea NetworkPolicy.
+	stageIngressSecurity
+	// Commit non-Service connections.
+	stageConntrack
+	// Output packets to target port.
+	stageOutput
+)
+
+// Table in FlexiblePipeline is the basic unit to build OVS pipelines. A Table can be referenced by one or more features,
+// but its member struct ofTable will be initialized and realized on OVS only when it is referenced by any activated features.
+type Table struct {
+	name     string
+	stage    binding.StageID // All stages are defined in pkg/ovs/openflow/interfaces.go
+	pipeline binding.PipelineID
+	ofTable  binding.Table
+}
+
+// tableOrderCache is used to save the order of all defined tables located in file pkg/agent/openflow/pipeline.go. The tables
+// have the same pipeline ID are saved in a slice by order of definition. When building a pipeline, the required table list
+// is aggregated from all activated features, and the order of the required tables is decided by the saved order.
+var tableOrderCache map[binding.PipelineID][]*Table
+
+// newTable is used to declare a Table. A table should belong to a stage and pipeline defined in current file.
+func newTable(tableName string, stage binding.StageID, pipeline binding.PipelineID) *Table {
+	table := &Table{
+		name:     tableName,
+		stage:    stage,
+		pipeline: pipeline,
+	}
+	if pipeline == pipelineRoot {
+		tableOrderCache = make(map[binding.PipelineID][]*Table)
+	} else {
+		tableOrderCache[pipeline] = append(tableOrderCache[pipeline], table)
+	}
+	return table
+}
+
+func (t *Table) GetID() uint8 {
+	return t.ofTable.GetID()
+}
+
+func (t *Table) GetNext() uint8 {
+	return t.ofTable.GetNext()
+}
+
+func (t *Table) GetName() string {
+	return t.name
+}
+
+func (t *Table) GetMissAction() binding.MissActionType {
```

Comment:
I don't see the method is used anywhere.

Reply:
It's used only once in function `setMockOFTables(ctrl *gomock.Controller, tableMap map[*Table]**mocks.MockTable)` of file `pkg/agent/openflow/network_policy_test.go`